13_BitonicSort/main.cpp

```diff
@@ -0,0 +1,319 @@
+			barrier1.barrier.dep.dstAccessMask = ACCESS_FLAGS::MEMORY_READ_BITS;
+			cmdbuf->pipelineBarrier(asset::E_DEPENDENCY_FLAGS(0), pipelineBarrierInfo1);
+
+			cmdbuf->pushConstants(m_pipeline->getLayout(), IShader::E_SHADER_STAGE::ESS_COMPUTE, 0u, sizeof(pc), &pc);
+
+			cmdbuf->dispatch(1, 1, 1);
+
+			IGPUCommandBuffer::SPipelineBarrierDependencyInfo pipelineBarrierInfo2 = {};
+			decltype(pipelineBarrierInfo2)::buffer_barrier_t barrier2 = {};
+			pipelineBarrierInfo2.bufBarriers = { &barrier2, 1u };
+			barrier2.range.buffer = m_deviceLocalBuffer;
+			barrier2.barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			barrier2.barrier.dep.srcAccessMask = ACCESS_FLAGS::MEMORY_WRITE_BITS;
+			barrier2.barrier.dep.dstStageMask = PIPELINE_STAGE_FLAGS::COPY_BIT;
+			barrier2.barrier.dep.dstAccessMask = ACCESS_FLAGS::MEMORY_READ_BITS;
+			cmdbuf->pipelineBarrier(asset::E_DEPENDENCY_FLAGS(0), pipelineBarrierInfo2);
+
+			copyInfo.srcOffset = 0;
+			copyInfo.dstOffset = outputOffset;
+			cmdbuf->copyBuffer(m_deviceLocalBuffer.get(), m_downStreamingBuffer->getBuffer(), 1, &copyInfo);
+			cmdbuf->end();
+		}
+
+		semaphoreValue++;
+		{
+			const IQueue::SSubmitInfo::SCommandBufferInfo cmdbufInfo =
+			{
+				.cmdbuf = cmdbuf.get()
+			};
+			const IQueue::SSubmitInfo::SSemaphoreInfo signalInfo =
+			{
+				.semaphore = m_timeline.get(),
+				.value = semaphoreValue,
+				.stageMask = asset::PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT
+			};
+
+			const IQueue::SSubmitInfo submitInfo = {
+				.waitSemaphores = {},
+				.commandBuffers = {&cmdbufInfo,1},
+				.signalSemaphores = {&signalInfo,1}
+			};
+
+			m_api->startCapture();
+			queue->submit({ &submitInfo,1 });
+			m_api->endCapture();
+		}
+
+		const ISemaphore::SWaitInfo futureWait = { m_timeline.get(),semaphoreValue };
+
+		m_upStreamingBuffer->multi_deallocate(AllocationCount, &inputOffset, &inputSize, futureWait);
+
+		auto latchedConsumer = make_smart_refctd_ptr<IUtilities::CDownstreamingDataConsumer>(
+			IDeviceMemoryAllocation::MemoryRange(outputOffset, outputSize),
+			[=](const size_t dstOffset, const void* bufSrc, const size_t size)->void
+			{
+				assert(dstOffset == 0 && size == outputSize);
+
+				std::cout << "Sorted array: ";
+				const uint32_t* const data = reinterpret_cast<const uint32_t*>(bufSrc);
+				for (auto i = 0u; i < elementCount; i++) {
+					uint32_t key = data[i * 2];
+					uint32_t value = data[i * 2 + 1];
+					std::cout << "(" << key << "," << value << "), ";
+					if ((i + 1) % 20 == 0) {
+						std::cout << "\n";
+					}
+				}
+				std::cout << "\nElement count: " << elementCount << "\n";
+
+				bool is_sorted = true;
+				int32_t error_index = -1;
+				for (uint32_t i = 1; i < elementCount; i++) {
+					uint32_t prevKey = data[(i - 1) * 2];
+					uint32_t currKey = data[i * 2];
+					if (currKey < prevKey) {
+						is_sorted = false;
+						error_index = i;
+						break;
+					}
```

Comment:
because Bitonic _should_ be stable, you should also check 
```cpp
else if (currKey==prevKey && currValue>prevValue) // check stability, this is why we've initialized the values in such a particular way
{
   // fail, not to bad sorting but instability
}
```

and also `assert(curValue!=prevValue);` because of how we initialized the values